ios/PurchasesHybridCommon/PurchasesHybridCommonSwift/CommonFunctionality.swift

```diff
@@ -0,0 +1,505 @@
+            if let error = error {
+                completion(nil, ErrorContainer(error: error, extraPayload: ["userCancelled": userCancelled]))
+            } else if let purchaserInfo = purchaserInfo,
+                      let transaction = transaction {
+                completion([
+                    "purchaserInfo": purchaserInfo.dictionary,
+                    "productIdentifier": transaction.payment.productIdentifier
+                ], nil)
+            } else {
+                // todo: maybe set up custom error here?
+                completion(nil, productNotFoundError(description: "Couldn't find product.", userCancelled: false))
+            }
+        }
+
+        product(with: productIdentifier) { product in
+            guard let product = product else {
+                completion(nil, productNotFoundError(description: "Couldn't find product.", userCancelled: false))
+                return
+            }
+
+            if let signedDiscountTimestamp = signedDiscountTimestamp {
+                if #available(iOS 12.2, macOS 10.14.4, tvOS 12.2, *) {
+                    guard let discount = self.discountsByProductIdentifier[signedDiscountTimestamp] else {
+                        completion(nil, productNotFoundError(description: "Couldn't find discount.", userCancelled: false))
+                        return
+                    }
+                    Purchases.shared.purchaseProduct(product, discount: discount, hybridCompletion)
+                    return
+                }
+
+                Purchases.shared.purchaseProduct(product, hybridCompletion)
+            }
+
+        }
+    }
+
+    @objc(purchasePackage:offering:signedDiscountTimestamp:completionBlock:)
+    static func purchasePackage(_ packageIdentifier: String,
+                                offeringIdentifier: String,
+                                signedDiscountTimestamp: String?,
+                                completion: @escaping ([String: Any]?, ErrorContainer?) -> Void) {
+        let hybridCompletion: (SKPaymentTransaction?,
+                               Purchases.PurchaserInfo?,
+                               Error?,
+                               Bool) -> Void = { transaction, purchaserInfo, error, userCancelled in
+            if let error = error {
+                completion(nil, ErrorContainer(error: error, extraPayload: ["userCancelled": userCancelled]))
+            } else if let purchaserInfo = purchaserInfo,
+                      let transaction = transaction {
+                completion([
+                    "purchaserInfo": purchaserInfo.dictionary,
+                    "productIdentifier": transaction.payment.productIdentifier
+                ], nil)
+            } else {
+                // todo: maybe set up custom error here?
+                completion(nil, productNotFoundError(description: "Couldn't find product.", userCancelled: false))
+            }
+        }
```

Comment:
I think this could be further cleaned up by extracting this bit. I didn't want to stray too far from the current implementation during migration, though, since we can always continue to clean up once we're fully migrated